service/header/sync.go

```diff
@@ -10,62 +11,119 @@ import (
 	tmbytes "github.com/tendermint/tendermint/libs/bytes"
 )
 
+// TODO:
+//  1. Sync protocol for peers to exchange their local heads on connect
+//  2. If we are far from peers but within an unbonding period - trigger sync automatically
+//  3. If we are beyond the unbonding period - request Local Head + 1 header from trusted and hardcoded peer
+//  automatically and continue sync until know head.
+//  4. Limit amount of requests on server side
+//  5. Sync status and till sync is done.
+//  6. Cache even unverifiable headers from the future, but don't resend them with ignore.
+//  7. Retry requesting headers
+
 // Syncer implements simplest possible synchronization for headers.
 type Syncer struct {
+	sub      *P2PSubscriber
 	exchange Exchange
 	store    Store
 	trusted  tmbytes.HexBytes
 
 	// inProgress is set to 1 once syncing commences and
 	// is set to 0 once syncing is either finished or
 	// not currently in progress
-	inProgress uint64
+	inProgress  uint64
+	triggerSync chan struct{}
+
+	// pending keeps ranges of valid headers rcvd from the network awaiting to be appended to store
+	pending *ranges
+
+	ctx    context.Context
+	cancel context.CancelFunc
 }
 
 // NewSyncer creates a new instance of Syncer.
-func NewSyncer(exchange Exchange, store Store, trusted tmbytes.HexBytes) *Syncer {
+func NewSyncer(exchange Exchange, store Store, sub *P2PSubscriber, trusted tmbytes.HexBytes) *Syncer {
 	return &Syncer{
-		exchange:   exchange,
-		store:      store,
-		trusted:    trusted,
-		inProgress: 0, // syncing is not currently in progress
+		sub:         sub,
+		exchange:    exchange,
+		store:       store,
+		trusted:     trusted,
+		triggerSync: make(chan struct{}),
+		pending:     newRanges(),
 	}
 }
 
 // Start starts the syncing routine.
 func (s *Syncer) Start(context.Context) error {
-	go s.Sync(context.TODO()) // TODO @Wondertan: leaving this to you to implement in disconnection toleration PR.
+	if s.sub != nil {
+		err := s.sub.AddValidator(s.validateMsg)
+		if err != nil {
+			return err
+		}
+	}
+
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	go s.syncLoop()
+	s.mustSync()
 	return nil
 }
 
+func (s *Syncer) wantSync() {
+	select {
+	case s.triggerSync <- struct{}{}:
+	default:
+	}
+}
+
+func (s *Syncer) mustSync() {
+	select {
+	case s.triggerSync <- struct{}{}:
+	case <-s.ctx.Done():
+	}
+}
+
+func (s *Syncer) syncLoop() {
+	for {
+		select {
+		case <-s.triggerSync:
+			s.sync(s.ctx)
+		case <-s.ctx.Done():
+			return
+		}
+	}
+}
+
 // Sync syncs all headers up to the latest known header in the network.
-func (s *Syncer) Sync(ctx context.Context) {
+func (s *Syncer) sync(ctx context.Context) {
 	log.Info("syncing headers")
 	// indicate syncing
 	s.syncInProgress()
 	// when method returns, toggle inProgress off
 	defer s.finishSync()
-	// TODO(@Wondertan): Retry logic
 	for {
-		localHead, err := s.getHead(ctx)
+		if ctx.Err() != nil {
+			return
+		}
```

Comment:
It's not immediately clear when this would happen, and if returning here is the right thing to do. Can you document this logic?

Reply:
Added a TODO explaining why it is needed.